File: app/src/main/java/eu/darken/sdmse/common/forensics/csi/dalvik/tools/SourceDirCheck.java
//package eu.darken.sdmse.common.forensics.csi.dalvik;
//
//
//import java.util.Collection;
//
//import eu.thedarken.sdm.tools.apps.SDMPkgInfo;
//import eu.thedarken.sdm.tools.forensics.Owner;
//import eu.thedarken.sdm.tools.forensics.OwnerInfo;
//import eu.thedarken.sdm.tools.forensics.csi.CSIHelper;
//import eu.thedarken.sdm.tools.forensics.csi.CSIProcessor;
//import eu.thedarken.sdm.tools.io.SDMFile;
//
//public class SourceDirCheck extends CSIHelper {
//
//    protected SourceDirCheck(CSIProcessor csiProcessor) {
//        super(csiProcessor);
//    }
//
//    public boolean check(OwnerInfo ownerInfo, Collection<SDMFile> candidates) {
//        for (SDMPkgInfo packageInfo : getApps()) {
//            if (packageInfo.getApplicationInfo() == null) continue;
//            for (SDMFile candidate : candidates) {
//                // Don't reverse the EQUALS because sourceDir can be null in some cases
//                if (candidate.getPath().equals(packageInfo.getSourceDir())) {
//                    ownerInfo.addOwner(new Owner(packageInfo.getPackageName()));
//                    return true;
//                }
//            }
//        }
//        return false;
//    }
//}

File: app/src/main/java/eu/darken/sdmse/common/forensics/csi/dalvik/tools/SourceDirCheck.kt
package eu.darken.sdmse.common.forensics.csi.dalvik.tools

import dagger.Reusable
import eu.darken.sdmse.common.files.local.LocalPath
import eu.darken.sdmse.common.forensics.AreaInfo
import eu.darken.sdmse.common.forensics.Owner
import eu.darken.sdmse.common.forensics.csi.dalvik.DalvikCheck
import eu.darken.sdmse.common.pkgs.PkgRepo
import eu.darken.sdmse.common.pkgs.currentPkgs
import javax.inject.Inject

@Reusable
class SourceDirCheck @Inject constructor(
    private val pkgRepo: PkgRepo,
) : DalvikCheck {

    suspend fun check(areaInfo: AreaInfo, candidates: Collection<LocalPath>): DalvikCheck.Result {
        val ownerPkg = pkgRepo.currentPkgs()
            .filter { it.sourceDir != null }
            .firstOrNull { pkg ->
                candidates.any { it.path == pkg.sourceDir?.path }

            }

        return DalvikCheck.Result(
            owners = setOfNotNull(ownerPkg?.id?.let { Owner(it, areaInfo.userHandle) })
        )
    }
}